src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,163 @@
+        */
+        T append(T newValue);
+        /**
+        * :return unordered content (data ) of this sl. window; call linearize() if you need them oredered from oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than the pure getData() but ensures that the data are ordered (oldest at the 
+         * beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> linearize() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+
+    // HELPER:
+     static std::vector<T> getLastNValues(std::vector<T> biggerData, nupic::UInt n) {
+      NTA_CHECK(n >=0 && n <= biggerData.size()); 
+    std::vector<T> v;
+    v.reserve(n);
+    v.insert(begin(v), end(biggerData) - n, end(biggerData));
+    NTA_ASSERT(v.size() == n);
+    return v;
+    };
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+        bool firstRun_ = true; //fill be false once "first run" (=size reaches maxCapacity for the first time) is completed
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::util;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  idxNext_ = 0;
+}
+
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(UInt maxCapacity, vector<T> initialData)  :
+  SlidingWindow(maxCapacity) {
+  NTA_CHECK(initialData.size() <= maxCapacity);
+  buffer_.insert(begin(buffer_), begin(initialData), end(initialData));
+  idxNext_ = initialData.size();
+}
+
+
+template<class T>
+UInt SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return buffer_.size();
+}
+
+
+template<class T>
+T SlidingWindow<T>::append(T newValue) {
+  Real old = NEUTRAL_VALUE;
+  if(firstRun_ && size() == maxCapacity) {
+    firstRun_ = false;
+  }
+
+  if(firstRun_) {
+    buffer_.emplace_back(newValue);
+  } else {
+    old = buffer_[idxNext_];  //FIXME this IF is here only because size() wouldn't work w/o it or similar hack
+    buffer_[idxNext_] = newValue;
+  }
+  idxNext_ = ++idxNext_ % maxCapacity;
+  return old;
+}
+
+
+template<class T>
+const vector<T>& SlidingWindow<T>::getData() const {
+  return buffer_;
+}
+
+
+template<class T>
+vector<T> SlidingWindow<T>::linearize() const {
+  vector<T> lin;
+  lin.reserve(buffer_.size());
+
+  if(size() ==0) return lin; //empty buffer
+  lin.insert(begin(lin), begin(buffer_) + idxNext_, end(buffer_)); //insert the "older" part at the beginning
```

Comment:
Would that be correct? We want to start writing to the `lin` buffer from the beginning.